utils/utils.c

```diff
@@ -58,8 +59,53 @@ void get_max_double(const int64_t ND1, const double *cz1, double *czmax)
     *czmax = max;
 }
 
+static int setup_bin_type_float(const float *rupp, int nbin, uint8_t *bin_type)
+{
+    // bin_type = 0: automatic detection
+    // bin_type = 1: linear
+    // bin_type = 2: custom
+    if (*bin_type == 0) {
+        // if linear spacing, return 1, else 0
+        const float atol = 1e-12;
+        const float rtol = 1e-12;
+        float rmin = rupp[0];
+        float rstep = (rupp[nbin-1] - rupp[0])/(nbin - 1);
+        *bin_type = 1;
+        for (int ii=1; ii<nbin; ii++) {
+            float pred = rmin + rstep*ii;
+            if ((fabsf(rupp[ii] - pred) > atol)||(fabsf((rupp[ii] - pred)/pred) > rtol)) {
+                *bin_type = 2;
+                break;
+            }
+        }
+    }
+    return EXIT_SUCCESS;
+}
+
+static int setup_bin_type_double(const double *rupp, int nbin, uint8_t *bin_type)
```

Comment:
As noted earlier, these functions should be available to call publicly (rather than static)